[WIP] Add hashing for types

phate opened the pull request:
What is still missing:
1. Provide documentation
2. Unit tests

phate commented:
@haved @caleridas I am not finished with it yet, but could I get a review to see whether this goes in the right direction, please?

caleridas commented on jlm/hls/ir/hls.cpp:
```diff
@@ -4,15 +4,40 @@
+std::size_t
+triggertype::ComputeHash() const noexcept
+{
+  return std::hash<const triggertype *>()(GetInstance());
```
I would suggest not hashing on some address, but simply using a constant magic value

Something like this would also work (it gets compile-time evaluated):

`constexpr std::size_t const_hash_str(std::string_view s) {
	std::size_t v = 0;
	for (auto c : s) {
		v = v * 5 + c;
	}
	return v;
}

std::size_t triggertype::ComputeHash() const noexcept { return constr_str_hash("triggertype"); }
`

phate replied:
I am fine with either, but why not hashing on a constant address? What is the rationale?

caleridas replied:
a) determinism b) compile-time constants are always nice I would say

phate replied:
Computing the hash at compile-time now where possible. I did not find a way to automatically get the qualified name of the type (`typeid(triggertype).name` did not do the trick), so I sadly had to provide manually the string.

haved reviewed (commented):
I have no comments beyond things I assume are already covered by //FIXMEs, or just me marveling at C++ features. 

I had to read up on shared pointers and fold expressions again to understand some of it, but it makes sense.

The only slight annoyance is that sometimes values are passed to `CombineHash`, and other times a hash is passed. In the latter case you end up hashing a hash, which isn't really a problem, but uses some extra CPU and might increase the chance of hash collisions by an extremely small margin.

phate commented on jlm/util/HashSet.hpp:
```diff
@@ -31,9 +31,18 @@ struct Hash<std::pair<First, Second>>
+void
+CombineHash(std::size_t & seed, const T & arg, Args... args)
+{
+  seed ^= std::hash<T>()(arg) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
```
Ultimately from boost (at some pointer).
Solution stolen from [here](https://stackoverflow.com/questions/2590677/how-do-i-combine-hash-values-in-c0x)
Explanation can be found [here](https://stackoverflow.com/questions/35985960/c-why-is-boosthash-combine-the-best-way-to-combine-hash-values)
I generally do not have a strong opinion here, but need something that just works without me re-inventing the wheel.

phate commented:
@haved I separated hash computation and combination from each other. This way we avoid taking the hash of a hash value.

caleridas commented on jlm/util/Hash.hpp:
```diff
@@ -0,0 +1,90 @@
+ * @return A hash value of \p s.
+ */
+constexpr std::size_t
+ComputeConstantHash(const std::string_view & s)
```
just pass std::string_view as value (no const &)

phate replied:
I removed the function again.

caleridas commented on jlm/util/Hash.hpp:
```diff
@@ -0,0 +1,90 @@
+ * Our own version of std::hash that also supports hashing std::pair
+ */
+template<typename T>
+struct Hash : std::hash<T>
```
TBH I am never sure which std:: classes you are allowed to derive from and which not. With rare exceptions, things in std:: should be treated as final. I am unclear about std::hash. Is hash-of-pair a recurring thing?

haved replied:
There is at least one place where HashSet of pair comes up, and I would think it is a common enough use case that it comes up again.

I'm noticing that the implementation currently does not support pairs of pairs, due to not using `util::Hash` in the implementation.

Regarding not inheriting from std::hash, I tried looking for information about it being illegal or discouraged. Since there is only one method, operator(), we could manually implement it and call `std::hash<T>()(value)` from there.

caleridas reviewed (approved):
Nothing much to say, does what it states that it does. Maybe name "ComputeHash" is a bit verbose... but does not really matter.

One thing that is going to be a footgun: `std::unordered_map<std::shared_ptr<const type>, ...>` might be something desirable to write, it even compiles and mostly works... except when it will not. I had considered introducing a wrapper type for std::shared_ptr<const type> that does not suffer from this problem, but I also do not like too many wrapping abstractions... not sure what is best.

haved commented on jlm/util/Hash.hpp:
```diff
@@ -0,0 +1,90 @@
+  std::size_t
+  operator()(const std::pair<First, Second> & value) const noexcept
+  {
+    return std::hash<First>()(value.first) ^ std::hash<Second>()(value.second) << 1;
```
If you want to you can replace `std::hash` with `Hash` to enable hashing pairs of pairs as well.

haved reviewed (approved):
Looks good, just one comment where I realize my original implementation could be better

phate commented:
@caleridas @haved I need a review again. The following things changed since last review:
1. I added unit tests for the `ComputeHash()` method for all type subclasses.
2. I realized that I need to add a subtype specific component to the hash, otherwise unrelated subtypes with the same attributes would produce the same hash. For example, `bittype` and `ctltype` might produce the same hash if no type component is added as both have only a size attribute (i.e., the number of bits and the number of alternatives) and if the value of both attributes was the same, then the hash was the same (for obvious reasons). The same problem occurred for `fixedvectortype`, `scalablevectortype`, and `arraytype`.
3. I realized that the C++ standard library provides [builtin hashes for types](https://en.cppreference.com/w/cpp/types/type_info/hash_code). I used this as the type specific component of the hash, avoiding all the home-grown strings. This in turn rendered the `ComputeConstantHash()` method superfluous and I removed it again. The problem is that the `hash_code()` is not computed at compilation time any longer. It seems to be the more reliable approach, but might not be as great in terms of determinism (quickly checking the result of `hash_code()` between program runs seems to result in the same hash value for `g++` and `clang++` though, but it does not seem to be guaranteed) and performance.
4. Changed `std::hash` to `Hash` for `std::pairs`.

haved commented on tests/jlm/hls/ir/TriggerTypeTests.cpp:
```diff
@@ -0,0 +1,30 @@
+  // Act & Assert
+  assert(triggerType1.ComputeHash() == triggerType1.ComputeHash());
+  assert(triggerType2.ComputeHash() == triggerType2.ComputeHash());
+  assert(triggerType1.ComputeHash() != ioStateType.ComputeHash());
```
It seems a bit off to assert that the hashes are strictly not the same, as hash collisions are possible. Since the hashes are based on `typeid().hash_code()` it might be the case that hash collisions are not possible (the standard recommends that implementations avoid this as much as possible)

phate replied:
I agree with you, but I have no idea how to better check it. There is a chance for hash collisions, but it should hopefully be very unlikely. Before I introduced the type aspect to the hash for each implementation, we had guaranteed collisions for hashes from different types (which should be very unlikely and not guaranteed). These checks ensure that we will hopefully not introduce a bug again such that it becomes guaranteed.

Honestly, I have no idea of how to improve upon this. @caleridas Any idea how to improve from here (if it is even worth it).

phate replied:
@haved @caleridas  I removed all the unit tests with the too strict asserts again.

haved reviewed (approved):
Looks good to me, my only concern is that the tests assert that hash collisions do not happen. With 64 bit hashes and using `typeid().hash_code()`, it should not be a problem, but the C++ standard does not strictly require different types to have different hash codes. In practice I think it is fine, as the asserts are only used in tests.

caleridas reviewed (approved):
looks good

haved reviewed (approved):
I agree, looks good
